src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
+      // memory.copy(dst, src, C)  ==>  store(dst, load(src))
+      if (auto* csize = memCopy->size->dynCast<Const>()) {
+        auto bytes = csize->value.geti32();
+
+        switch (bytes) {
+          case 0: {
+            if (!destHasSideEffects &&
+                !EffectAnalyzer(getPassOptions(), features, memCopy->source)
+                   .hasSideEffects()) {
+              return ExpressionManipulator::nop(curr);
+            }
+            break;
+          }
+          case 1:
+          case 2:
+          case 4: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
+              memCopy->dest,
+              builder.makeLoad(
+                bytes, false, 0, bytes, memCopy->source, Type::i32),
+              Type::i32);
+          }
+          case 8: {
+            Builder builder(*getModule());
```

Comment:
You could reduce repetition by constructing the `builder` before the switch.

Reply:
Thanks! Basically at first time I implemented in exactly that way, but I decided put builder init inside cases due to it more lazily initialization and could be avoided in most of cases (when count of bytes is not pow of two). But I could revert this